### PR TITLE
IPQ 03: Array-to-BST

### PR DESCRIPTION
**Pull request by char-adadev**
The purpose of this PR is to introduce the Array-to-BST IPQ. This PR builds on the PR for Linked List Intersection.

For review:
* Please run learn preview
* Verify the example implementation passes tests
* Any suggestions for wording, grammar, etc are welcome

### CODE REVIEW

File: 03/arr-to-bst.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+![Balanced Binary Search Tree](../images/balanced_bst.png)
+
+Please note one is not required to implement a self-balancing Binary Search Tree in order to solve this exercise. 
+
+It is recommended to break the problem down recursively by first setting the root of the Binary Search Tree to the middle element of the array.
+
+<br>
+<details style="max-width: 700px; margin: auto;">
+<summary>Click here to see the tests that will be run against your code</summary>
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:21]: Again, move these below the implementation question!

File: 03/arr-to-bst.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+`arr = [5, 10, 15, 20, 25, 30, 35, 40, 45]`
+
+should result in a tree with the following root/height:
+
+![Balanced Binary Search Tree](../images/balanced_bst.png)
+
+Please note one is not required to implement a self-balancing Binary Search Tree in order to solve this exercise. 
+
+It is recommended to break the problem down recursively by first setting the root of the Binary Search Tree to the middle element of the array.
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:17]: Consider moving this to the hints section of the sample pseudocode implmentation

File: 03/arr-to-bst.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+For each question, provide an explanation which includes the effect your decision would have on how you would approach the problem.
+
+##### !end-question
+
+##### !explanation
+
+Here are some example clarifying questions:
+
+1. What kind of data is stored in the Tree Nodes?
+2. Is the array guaranteed to be non-empty?
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:97]: Consider adding 'Can there be duplicate elements in the array?' Or perhaps we should address this in the problem statement.

File: 03/arr-to-bst.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+* topics: pse
+##### !question
+
+1. Use the comments provided to write at least two example input/outputs:
+    * Consider at least one nominal and one edge case.
+    * What is the expected output for the given input?
+    * You can use the examples provided in the prompt, or other examples.
+2. Write unit tests for `arr_to_bst` for the nominal and edge cases you identified in the first step.
+
+*Note: Click the **Run Tests** button to save your tests for instructor feedback. No real tests are actually run again your unit tests.*
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:121]: Hmm... should we provide them with `is_bst`, `height` and `is_balanced_tree` here? We may need to rethink how we frame the question for this IPQ since the tests are a bit more involved.

File: 03/arr-to-bst.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        # Arrange
+        arr = [5, 10, 15, 20, 25, 30, 35, 40, 45]
+
+        # Act
+        result = arr_to_bst(arr)
+
+        # Assert
+        self.assertEqual(result.val, 25)
+        self.assertTrue(is_bst(result))
+        self.assertTrue(is_balanced_tree(result))
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:182]: Students are taught pytest but not unittest in the classroom phase. While I would expect them to be able to read and understand a unittest, I'm not sure writing them should be within scope for this course. 

```suggestion
        assert result.val == 25
        assert is_bst(result)
        assert is_balanced_tree(result)
```

File: 03/arr-to-bst.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+def test_will_return_none_for_empty_list(self):
+        # Arrange
+        arr = []
+
+        # Act
+        result = arr_to_bst(arr)
+
+        # Assert
+        self.assertEqual(result, None)
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:192]:
```suggestion
        assert result is None
```

File: 03/arr-to-bst.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+Example:
+
+`arr = [5, 10, 15, 20, 25, 30, 35, 40, 45]`
+
+should result in a tree with the following root/height:
+
+![Balanced Binary Search Tree](../images/balanced_bst.png)
+
+Please note one is not required to implement a self-balancing Binary Search Tree in order to solve this exercise.
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:15]:
```suggestion
Please note one is not required to implement a self-balancing Binary Search Tree in order to solve this exercise. For an extra challenge, think about why it's unnecessary!
```

**Review by kyra-at-ada — COMMENTED** [review]
This looks great! I think we should discuss how we want to ask folks to write test cases because I wouldn't expect them to write helper functions to assess if an object is a valid bst or check if a bst is balanced as part of this exercise. Do we want to change it to a pseudocode? We could prepend the helper functions (although it gets a bit long). We could write function stubs with docstrings so Learn doesn't throw an error when they run their code.

File: 03/arr-to-bst.md (hunk 6 above)
Comment (char-adadev) [reply 03/arr-to-bst.md]: Updated!

File: 03/arr-to-bst.md (hunk 5 above)
Comment (char-adadev) [reply 03/arr-to-bst.md]: updated 🙌🏾

File: 03/arr-to-bst.md (hunk 4 above)
Comment (char-adadev) [reply 03/arr-to-bst.md:65]: I added a note that `is_bst` and `is_balanced_tree` functions are available to them. I don't believe the tests they write here are actually ever run (verified by going into Learn and writing junk code in here). Since the Test Suite only runs the `test_always_pass` test, anything they write in here should be able to run.

That being said, I would love for them to think about how these tests are written, so I also added the `is_bst`, `height`, and `is_balanced_tree` functions to the example tests. I'm hopeful that'll make it super clear what is happening in those example tests.

Lemme know if anything seems amiss!

File: 03/arr-to-bst.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        self.assertEqual(1,1)
+```
+
+##### !end-tests
+##### !explanation 
+
+Example tests:
+
+```python
+def test_will_return_balanced_bst_for_odd_lengthed_list(self):
```
Comment (kyra-at-ada) [inline 03/arr-to-bst.md:167]: We should probably remove `self` in all the examples as that's also a carry over from `unittest`

File: 03/implement-arr-to-bst.checkpoint.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<br>
+<details style="max-width: 700px; margin: auto;">
+<summary>Click here to see the tests that will be run against your code</summary>
+
+```py
+def test_will_return_balanced_bst_for_odd_lengthed_list(self):
```
Comment (kyra-at-ada) [inline 03/implement-arr-to-bst.checkpoint.md:184]: Same here regarding `self` in the example tests. Or since it looks like this still uses `unittest` maybe add a note about what `self` is referencing?

**Review by kyra-at-ada — APPROVED** [review]
Looks great! I think you added info about the helper functions for testing in a really concise way. 

Added a couple of comments about referencing `self` with testing, but it looks pretty much ready to go!
